incuna_auth/middleware/login_required.py

```diff
@@ -38,40 +36,35 @@ class LoginRequiredMiddleware:
     Will default to protecting everything if LOGIN_PROTECTED_URLS is not in
     settings.
 
-
     Original code from:
     http://onecreativeblog.com/post/59051248/django-login-required-middleware
 
     This version has been modified to allow us to define areas of the site to
     password protect instead of protecting everything under /.
     '''
-    def process_request(self, request):
-        assert hasattr(request, 'user'), (
-            "The 'LoginRequiredMiddleware' "
-            + "requires authentication middleware to be installed. Edit your "
-            + "'MIDDLEWARE_CLASSES' setting to insert "
-            + "'django.contrib.auth.middlware.AuthenticationMiddleware'. "
-            + "If that doesn't work, ensure your 'TEMPLATE_CONTEXT_PROCESSORS' "
-            + "setting includes 'django.core.context_processors.auth'."
-        )
-
-        # Jump over this middleware if not a protected url.
-        path = request.path_info.lstrip('/')
-        path_is_exempt = any(m.match(path) for m in EXEMPT_URLS)
-        path_is_protected = any(m.match(path) for m in PROTECTED_URLS)
-        if path_is_exempt or not path_is_protected:
-            return
-
-        # Jump over this middleware if user logged in.
-        if request.user.is_authenticated():
-            return
+    """
+    login_exempt_urls = getattr(settings, 'LOGIN_EXEMPT_URLS', [])
+    login_protected_urls = getattr(settings, 'LOGIN_PROTECTED_URLS', [r'^'])
 
-        # Raise a 403 for POST/DELETE etc.
-        if request.method != 'GET':
-            return HttpResponseForbidden()
+    EXEMPT_URLS = [compile_url(settings.LOGIN_URL), compile_url(settings.LOGOUT_URL)]
+    EXEMPT_URLS += compile_urls(login_exempt_urls)
+    PROTECTED_URLS = compile_urls(login_protected_urls)
 
-        # Add a message, and redirect to login.
-        if SEND_MESSAGE:
-            messages.info(request, _('You must be logged in to view this page.'))
+    def process_request(self, request):
+        self.assert_request_has_user(request)
+
+        # Jump over this middleware if the page doesn't come under its jurisdiction.
+        if not self.is_url_protected(
+            request,
+            feincms_states=[AccessState.STATE_AUTH_ONLY],
+            exempt_urls=self.EXEMPT_URLS,
+            protected_urls=self.PROTECTED_URLS
+        ):
+            return
 
-        return redirect_to_login(request.path_info)
+        # Deny access if the user is not logged in.
+        if request.user.is_anonymous():
+            error = ''
+            if SEND_MESSAGE:
+                error = 'You must be logged in to view this page.'
+            return self.deny_access(request, error)
```

Comment:
I may or may not understand what you're saying.  Are you suggesting doing this?

```python
class MiddlewareMixin:
    ...

    def process_request(self):
        if not self.is_url_protected(args):
            return

        if self.deny_access_condition():
            self.deny_access(self.message)
```

I thought about that earlier actually, but I don't like it much.  It's too prescriptive.  Even `LoginRequiredMiddleware` itself has a bit more logic in `process_request` (it calls `assert_has_user`, which not everything will need) and switches whether or not to send a Django message based on a local setting. (I could move the latter bit of logic into `MiddlewareMixin`, but that wouldn't help either - once you have multiple middlewares in a project, each one might require its own setting to switch messaging on and off.)